File: WebChat.Domain/Entities/Group.cs
namespace WebChat.Domain.Entities;

public class Group : AuditableEntity, IAggregateRoot
{
    public Guid GroupId { get; set; } = Guid.Empty;

    public string? AvatarPath { get; set; } = default;

    public string Name { get; set; } = string.Empty;

    public int NumberOfMembers => Members.Count;

    public virtual ICollection<JoinRequest> JoinRequests { get; set; } = new List<JoinRequest>();

    public virtual ICollection<JoinInvitation> JoinInvitations { get; set; } = new List<JoinInvitation>();

    public virtual ICollection<GroupMember> Members { get; set; } = new List<GroupMember>();

    public virtual GroupSetting Setting { get; set; } = new();

    public Guid ConversationId { get; set; } = Guid.Empty;

    public override Guid GetId() => GroupId;
}

public class GroupMember
{
    public Guid GroupId { get; set; } = Guid.Empty;

    public Guid UserId { get; set; } = Guid.Empty;

    public MemberRole Role { get; set; } = MemberRole.Member;

    public DateTime JoinDate { get; set; } = DateTime.UtcNow;

    public string? JoinBy { get; set; }
}

public class JoinRequest
{
    public Guid GroupId { get; set; } = Guid.Empty;

    public Guid UserId { get; set; } = Guid.Empty;

    public DateTime SendTime { get; set; } = DateTime.UtcNow;
}

public class JoinInvitation
{
    public Guid GroupId { get; set; } = Guid.Empty;

    public Guid UserId { get; set; } = Guid.Empty;

    public Guid MemberId { get; set; } = Guid.Empty;

    public DateTime SendTime { get; set; } = DateTime.UtcNow;
}

public class GroupSetting
{
    public bool AllowChangeGName { get; set; } = true;

    public bool AllowSendGInvitation { get; set; } = true;

    public bool MembershipApproval { get; set; } = false;

    public bool ReadRecentMessage { get; set; } = true;

    public bool JoinGroupByLink { get; set; } = true;

    public string GroupCode { get; set; } = string.Empty;
}

File: WebChat.Client/src/types/Group.ts
import { MemberRole } from "../enums/MemberRole";

export interface JoinInvitation {
  groupId: string;
  userId: string;
  memberId: string;
  sendTime: Date;
}

export interface JoinRequest {
  groupId: string;
  userId: string;
  sendTime: Date;
}

export interface GroupMember {
  groupId: string;
  userId: string;
  role: MemberRole;
  joinDate: Date;
  joinBy?: string | null;
}

export interface GroupSetting {
  allowChangeGName: boolean;
  allowSendGInvitation: boolean;
  membershipApproval: boolean;
  readRecentMessage: boolean;
  joinGroupByLink: boolean;
  groupCode: string;
}

export default interface Group {
  groupId: string;
  avatar: boolean;
  name: string;
  numberOfMembers: number;
  joinRequests: Array<JoinRequest>;
  joinInvitations: Array<JoinInvitation>;
  members: Array<GroupMember>;
  setting: GroupSetting;
  conversationId: string;
}
